File: Q1162 - As Far from Land as Possible/1162.cpp
class Solution
{
public:
  int maxDistance(vector<vector<int>> &grid)
  {
    int row = grid.size();
    int column = grid[0].size();

    int Y = row - 1;
    int X = column - 1;

    vector<vector<int>> values(row, vector<int>(column, INT_MAX));

    queue<pair<int, int>> queue;
    for (int y = 0; y != row; y++)
    {
      for (int x = 0; x != column; x++)
      {
        if (grid[y][x] == 0 && ((y && grid[y - 1][x]) || (y != Y && grid[y + 1][x]) || (x && grid[y][x - 1]) || (x != X && grid[y][x + 1])))
        {
          queue.push({y, x});
          values[y][x] = 1;
        }
      }
    }
    if (queue.empty())
    {
      return -1;
    }
    int w = 0;
    while (!queue.empty())
    {
      w++;
      int n = queue.size();
      while (n--)
      {
        auto [y, x] = queue.front();
        queue.pop();

        if (y && grid[y - 1][x] == 0 && values[y - 1][x] > w)
        {
          queue.push({y - 1, x});
          values[y - 1][x] = w;
        }
        if (y != Y && grid[y + 1][x] == 0 && values[y + 1][x] > w)
        {
          queue.push({y + 1, x});
          values[y + 1][x] = w;
        }

        if (x && grid[y][x - 1] == 0 && values[y][x - 1] > w)
        {
          queue.push({y, x - 1});
          values[y][x - 1] = w;
        }
        if (x != X && grid[y][x + 1] == 0 && values[y][x + 1] > w)
        {
          queue.push({y, x + 1});
          values[y][x + 1] = w;
        }
      }
    }
    return w;
  }
};

File: Q1162 - As Far from Land as Possible/1162.py
class Solution:
    def maxDistance(self, grid: List[List[int]]) -> int:

        row = len(grid)
        q = []
        dp = [[-1 for _ in range(row)] for _ in range(row)]

        def isvalid(i,j):
            if (0<=i<row and 0<=j<row and grid[i][j]==0):
                return True
            return False

        for i in range(row):
            for j in range(row):
                if (grid[i][j]==1):
                    q.append((i,j))
                    dp[i][j]=0

        res = -1
        while q:
            x,y = q.pop(0)
            for dx,dy in [(1,0),(-1,0),(0,-1),(0,1)]:
                if (isvalid(x+dx,y+dy) and dp[x+dx][y+dy]==-1):
                    q.append((x+dx,y+dy))
                    dp[x+dx][y+dy] = dp[x][y]+1
                    res = max(res,dp[x+dx][y+dy])

        return res